### ISSUE TITLE
Integrators should provide custom explanations why they are requesting scopes from the user

### ISSUE THREAD

**Issue by harrisj** [18F/myusa#675]
In the [new wireframes for login flow](https://projects.invisionapp.com/share/D338FOU9G#/screens), each scope requested by Integrators has an accompanying _Why?_ link that pops up a short explanation of why the integrator needs that field. There is currently no place in the backend for storing this information. Adding this would involve the following steps:

* [ ] Write a migration to add the information to the database. If we match the current schema, this would mean adding `explanations` as a single delimited text field to match the `scopes` column in `oauth_applications` (rather than a separate `has_many` association)
* [ ] method to return scopes as an array of hashes with the explanations and names, rather than just the names where used by the front-end rendering code
* [ ] modifications to the developer admin that will require customers to write out explanation text for each scope they select for their application. It might be good to give them some basic examples for each field.

We should have a discussion about schema. It's arguably cleaner to extract the scope selections to a separate database table named `oauth_application_scopes` that is `has_many` to `oauth_applications` and replaces the current `scopes` and `explanations` columns that would be in that table. This would make it much cleaner to add other attributes around scopes like `required?` down the road without keeping the contents of several database columns in sync.

**Comment by adelevie** [comment]
We should also have default values for these, right?

**Comment by harrisj** [comment]
Agreed. That would mean an additional migration to the `oauth_scopes` table to add a default explanation or whatever is the proper name for one of these fields.

**Comment by adelevie** [comment]
Bonus points if these are editable via API.

**Comment by esgoodman** [comment]
I'd settle for editable by us for now! But, honestly, this is a change to
the normal permissions help text paradigm for both integrators and users.
So I'm expecting it'll take a few iterations to figure out how  it should
work.

**Comment by harrisj** [comment]
To be honest, it's an infrequent enough thing to change that it would not
necessarily be a good API requirement. But maybe down the road.

Sent from my iPhone

On Jun 15, 2015, at 8:13 PM, Elizabeth Goodman <notifications@github.com>
wrote:

I'd settle for editable by us for now! But, honestly, this is a change to
the normal permissions help text paradigm for both integrators and users.
So I'm expecting it'll take a few iterations to figure out how it should
work.

—
Reply to this email directly or view it on GitHub
<https://github.com/18F/myusa/issues/675#issuecomment-112242242>.

**Comment by harrisj** [comment]
One other consideration: the `oauth_applications` table is created by a migration within the Devise plugin, so it's probably ill-advised to significantly alter that table's schema as I proposed would be better.

**Comment by harrisj** [comment]
The approach using a separate application_scopes model seems more viable and only a little bit of emulation of the `Doorkeeper::Models::Scopes` mixin methods. The only thing left to do is the input form for integrators to specify the reasons.
